daemon/algod/api/server/v2/dryrun.go

```diff
@@ -0,0 +1,512 @@
+	ddr.history = append(ddr.history, st)
+	ddr.updateScratch()
+	return nil
+}
+
+// Complete is called when the program exits (DebuggerHook interface)
+func (ddr *dryrunDebugReceiver) Complete(state *logic.DebugState) error {
+	return ddr.Update(state)
+}
+
+type dryrunLedger struct {
+	// inputs:
+
+	dr    *DryrunRequest
+	proto *config.ConsensusParams
+
+	// intermediate state:
+
+	// index into dr.Accounts[]
+	accountsIn map[basics.Address]int
+
+	// index into dr.Apps[]
+	accountApps map[basics.Address]int
+
+	// accounts that have been Put
+	accounts map[basics.Address]basics.BalanceRecord
+}
+
+func (dl *dryrunLedger) init() error {
+	dl.accounts = make(map[basics.Address]basics.BalanceRecord)
+	dl.accountsIn = make(map[basics.Address]int)
+	dl.accountApps = make(map[basics.Address]int)
+	for i, acct := range dl.dr.Accounts {
+		xaddr, err := basics.UnmarshalChecksumAddress(acct.Address)
+		if err != nil {
+			return err
+		}
+		dl.accountsIn[xaddr] = i
+	}
+	for i, app := range dl.dr.Apps {
+		var addr basics.Address
+		if app.Creator != "" {
+			var err error
+			addr, err = basics.UnmarshalChecksumAddress(app.Creator)
+			if err != nil {
+				return err
+			}
+		}
+		dl.accountApps[addr] = i
+	}
+	return nil
+}
+
+// transactions.Balances interface
+func (dl *dryrunLedger) Round() basics.Round {
+	return basics.Round(dl.dr.Round)
+}
+
+// transactions.Balances interface
+func (dl *dryrunLedger) Get(addr basics.Address, withPendingRewards bool) (basics.BalanceRecord, error) {
+	// first check accounts from a previous Put()
+	br, ok := dl.accounts[addr]
+	if ok {
+		return br, nil
+	}
+	// check accounts from debug records uploaded
+	any := false
+	out := basics.BalanceRecord{
+		Addr: addr,
+	}
+	accti, ok := dl.accountsIn[addr]
+	if ok {
+		any = true
+		acct := dl.dr.Accounts[accti]
+		var err error
+		if out.AccountData, err = AccountToAccountData(&acct); err != nil {
+			return basics.BalanceRecord{}, err
+		}
+		if withPendingRewards {
+			out.MicroAlgos.Raw = acct.Amount
+		} else {
+			out.MicroAlgos.Raw = acct.AmountWithoutPendingRewards
+		}
+	}
+	appi, ok := dl.accountApps[addr]
+	if ok {
+		any = true
+		app := dl.dr.Apps[appi]
+		out.AppParams = make(map[basics.AppIndex]basics.AppParams)
+		out.AppParams[basics.AppIndex(app.AppIndex)] = ApplicationParamsToAppParams(&app.Params)
+	}
+	if !any {
+		return basics.BalanceRecord{}, fmt.Errorf("no account for addr %s", addr.String())
+	}
+	return out, nil
+}
+
+// transactions.Balances interface
+func (dl *dryrunLedger) Put(br basics.BalanceRecord) error {
+	if dl.accounts == nil {
+		dl.accounts = make(map[basics.Address]basics.BalanceRecord)
+	}
+	dl.accounts[br.Addr] = br
+	return nil
+}
+
+// PutWithCreatables is like Put, but should be used when creating or deleting an asset or application.
+func (dl *dryrunLedger) PutWithCreatables(record basics.BalanceRecord, newCreatables []basics.CreatableLocator, deletedCreatables []basics.CreatableLocator) error {
+	return nil
```

Comment:
is that a placeholder ?

Reply:
I think this is here because `dryrunLedger` needs to fulfill the `Balances` interface. @algorandskiy can you confirm?